rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,708 @@
+
+### Train departing 15 minutes later than scheduled
+- Inspector Alice (badge number: 123) is working at Mattapan
+- The 1:30am trip is being held 15 minutes to wait for a final Red Line train
+```json
+{
+  "type": "com.mbta.ctd.glides.trips_updated.v1",
+  "time": "2023-01-23T01:25:00-05:00",
+  "data": {
+    "metadata": {
+      "inputType": "edit-trip"
+    },
+    "tripUpdates": [
+      {
+        "type": "updated",
+        "tripKey": {
+          "serviceDate": "2023-01-22",
+          "startLocation": {"gtfsId": "place-matt"},
+          "endLocation": {"gtfsId": "place-ashmt"},
+          "startTime": "25:30:00",
+          "endTime": "25:38:00"
+        },
+        "location": {"gtfsId": "place-matt"},
+        "startTime": "25:45:00",
+        "scheduled": {"scheduledCars": [{}]}
+      }
+    ]
+  }
+}
+```
+
+Note that the date of the the event is 2023-01-23, but the service date is 2023-01-22. The hour in the time itself is greater than 24 hours, to reflect that the time is the next calendar day.
+
+# Drawbacks
+- This requires Glides to include new functionality to write their events into a Kinesis stream. However, they already use `ex_aws` so this additional functionality should not be hard to add.
+ 
+# Rationale and alternatives
+- RTR already consumes a Kinesis stream of CloudEvents from OCS, so it's a limited additional lift to consume a new Kinesis stream.
+- Kinesis only adds a limited (<1s) latency between when the event is generated and when it is available to consumers.
+- Kinesis Firehose allows writing the stream of records to S3 for LAMP to consume, without needing a separate integration.
+
+## Alternative: allow setting times at non-terminal locations
+An earlier version of this version supported setting arrival/departure times for non-terminal locations as a part of the [Trip updated](#trip-updated) event. This is not supported by Glides, and there are no plans to support it: see [Future possibilities](#future-possibilities).
+
+## Alternative: allow `location_type` 0 locations
+While more flexible, it requires consumers of the events to support two types of GTFS stops. Glides has no plans to use the additional flexibility.
+
+## Alternative: events for importing the schedule
+Instead of the `scheduledOperators` field in [Trip updated](#trip-updated), another possibility would be including events which allow consumers to determine the scheduled operator. However, these events are complicated to implement, and require consumers to maintain state for many months (as the schedule events would not be generated frequently).
+
+## Alternative: times as RFC3999 string
+This would be easy to parse with standard tools, but doesn't match existing OCS behavior which only provides a time.
```

Comment:
thought: This time change is going to require new code in RTR either way since OCS doesn't use GTFS times: it uses 24-hour times that reset at midnight.

Reply:
It's not significantly different for the Glides code to publish either. Which do you think would be better?